lib/Reader/abisignature.cpp

```diff
@@ -86,43 +105,219 @@ ABICallSignature::ABICallSignature(const ReaderCallSignature &TheSignature,
                                    GenIR &Reader, const ABIInfo &TheABIInfo)
     : ABISignature(TheSignature, Reader, TheABIInfo), Signature(TheSignature) {}
 
-Value *ABICallSignature::emitCall(GenIR &Reader, llvm::Value *Target,
-                                  bool MayThrow, llvm::ArrayRef<Value *> Args,
-                                  llvm::Value *IndirectionCell,
-                                  llvm::Value **CallNode) const {
+static Value *getFieldAddress(IRBuilder<> &Builder, Value *Base,
+                              uint32_t Offset, Type *FieldTy) {
+  // The base value should be an i8*.
+  assert(Base->getType()->isPointerTy());
+  assert(Base->getType()->getPointerElementType()->isIntegerTy(8) ||
+         Base->getType()
+             ->getPointerElementType()
+             ->getArrayElementType()
+             ->isIntegerTy(8));
+
+  Type *Int32Ty = Type::getInt32Ty(Builder.getContext());
+  Value *Indices[] = {ConstantInt::get(Int32Ty, 0),
+                      ConstantInt::get(Int32Ty, Offset)};
+  Value *Address = Builder.CreateInBoundsGEP(Base, Indices);
+  PointerType *AddressTy = cast<PointerType>(Address->getType());
+  if (AddressTy->getElementType() != FieldTy) {
+    AddressTy = PointerType::get(FieldTy, AddressTy->getAddressSpace());
+    Address = Builder.CreatePointerCast(Address, AddressTy);
+  }
+  return Address;
+}
+
+CallSite ABICallSignature::emitUnmanagedCall(GenIR &Reader, Value *Target,
+                                             bool MayThrow,
+                                             ArrayRef<Value *> Arguments,
+                                             Value *&Result) const {
+  const LLILCJitContext &JitContext = *Reader.JitContext;
+  const struct CORINFO_EE_INFO::InlinedCallFrameInfo &CallFrameInfo =
+      JitContext.EEInfo.inlinedCallFrameInfo;
+  LLVMContext &LLVMContext = *JitContext.LLVMContext;
+  Type *Int8Ty = Type::getInt8Ty(LLVMContext);
+  Type *Int32Ty = Type::getInt32Ty(LLVMContext);
+  Type *Int8PtrTy = Reader.getUnmanagedPointerType(Int8Ty);
+  IRBuilder<> &Builder = *Reader.LLVMBuilder;
+
+  Reader.insertIRForUnmanagedCallFrame();
+  assert(Reader.UnmanagedCallFrame != nullptr);
+  assert(Reader.ThreadPointer != nullptr);
+
+  Value *CallFrame = Reader.UnmanagedCallFrame;
+  Value *Thread = Reader.ThreadPointer;
+
+  // Set m_pDatum if necessary
+  if (Reader.MethodSignature.hasSecretParameter()) {
+    // TODO: this needs to be updated for direct unmanaged calls.
```

Comment:
Direct unmanaged calls require the method handle instead of the stub secret parameter. This will require some additional plumbing (probably nothing more than a new field on ReaderCallSignature).